source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+      // loop variant.
+      if (loop->IsInsideLoop(unknown->UniqueId())) return false;
+    }
+  }
+
+  return true;
+}
+
+std::string SENode::AsString() const {
+  switch (GetType()) {
+    case Constant:
+      return "Constant";
+    case RecurrentExpr:
+      return "RecurrentExpr";
+    case Add:
+      return "Add";
+    case Negative:
+      return "Negative";
+    case Multiply:
+      return "Multiply";
+    case ValueUnknown:
+      return "Value Unknown";
+    case CanNotCompute:
+      return "Can not compute";
+  }
+  return "NULL";
+}
+
+bool SENode::operator==(const SENode& other) const {
+  if (GetType() != other.GetType()) return false;
+
+  if (other.GetChildren().size() != children_.size()) return false;
+
+  for (size_t index = 0; index < children_.size(); ++index) {
+    if (other.GetChildren()[index] != children_[index]) return false;
+  }
+
+  // If we're dealing with a value unknown node check both nodes were created by
+  // the same instruction.
+  if (GetType() == SENode::ValueUnknown) {
+    if (AsSEValueUnknown()->UniqueId() !=
+        other.AsSEValueUnknown()->UniqueId()) {
+      return false;
+    }
+  }
+
+  if (AsSEConstantNode()) {
+    if (AsSEConstantNode()->FoldToSingleValue() !=
+        other.AsSEConstantNode()->FoldToSingleValue())
+      return false;
+  }
+
+  return true;
+}
+
+bool SENode::operator!=(const SENode& other) const { return !(*this == other); }
+
+// Implements the hashing of SENodes.
+size_t SENodeHash::operator()(const SENode* node) const {
+  // Hashing the type as a string is safer than hashing the enum as the enum is
+  // very likely to collide with constants.
+  std::string type = node->AsString();
+
+  // We just ignore the literal value unless it is a constant.
+  int64_t literal_value = 0;
+  if (node->GetType() == SENode::Constant)
+    literal_value = node->AsSEConstantNode()->FoldToSingleValue();
+
+  // Hash the type string and the constant value if any.
+  size_t resulting_hash =
+      std::hash<std::string>{}(type) ^ std::hash<int64_t>{}(literal_value);
+
+  // If we're dealing with a recurrent expression hash the loop as well so that
+  // nested inductions like i=0,i++ and j=0,j++ correspond to different nodes.
+  if (node->GetType() == SENode::RecurrentExpr) {
+    resulting_hash ^=
+        std::hash<const ir::Loop*>{}(node->AsSERecurrentNode()->GetLoop());
+  }
+
+  // Hash the unique id of the creator instruction if this is a value unknown
+  // node.
+  if (node->GetType() == SENode::ValueUnknown) {
+    resulting_hash ^=
+        std::hash<uint32_t>{}(node->AsSEValueUnknown()->UniqueId());
+  }
+  // Hash the pointers of the child nodes, each SENode has a unique pointer
+  // associated with it.
+  const std::vector<SENode*>& children = node->GetChildren();
+  for (const SENode* child : children) {
+    resulting_hash ^= std::hash<const SENode*>{}(child);
+  }
+
+  return resulting_hash;
```

Comment:
The xor leads to a weak hash. Look at redundancy elimination or the type manager for better examples.